### PR TITLE
Feature and API improvements to Atracsys device

### PR DESCRIPTION
**Pull request by markasselin**
@hgueziri Could you please give this a quick try with the hardware and let me know if all is working? Unfortunately I don't have easy access to an Atracsys system at the moment.

@lassoan @adamrankin I'm running a test build right now, if it succeeds this should be ready to integrate. Will update when it completes.

### CODE REVIEW

**Comment by markasselin** [comment]
@lassoan @adamrankin This builds successfully with the latest Visual Studio on Windows. Ready to integrate!

It would be helpful if someone could test with the hardware again and let me know if there are any bugs, but I have done a complete test of this code some months ago on both spryTrack & fusionTrack devices.

File: src/PlusDataCollection/Atracsys/vtkPlusAtracsysTracker.cxx [hunk 1]
```diff
@@ -10,11 +10,13 @@ See License.txt for details.
...
 #include "vtkIGSIOAccurateTimer.h"
 #include "vtkPlusAtracsysTracker.h"
+#include "vtkPlusDataSource.h"
 
 // VTK includes
 #include <vtkMath.h>
-#include <vtkNew.h>
 #include <vtkMatrix4x4.h>
+#include <vtkNew.h>
+#include <vtkSmartPointer.h>
```
Comment (adamrankin) [inline src/PlusDataCollection/Atracsys/vtkPlusAtracsysTracker.cxx]: Yeah, someone else who sorts by alphabetical!!

**Comment by hgueziri** [comment]
@markasselin sorry for the delay I couldn't try it yesterday.

It compiles fine on Linux, but the server is not able to connect to the device anymore, at least with the fusion track 500.

Here is the output with verbose=5

```
System start timestamp: 1576079838
Software version: Plus-2.9.0.5bb19899 - Linux
|INFO|000.000659| Logging at level 5 (TRACE) to file: /home/hgueziri/Plus4Ibis/PlusBuild-bin/bin/Output/121119_105717_PlusLog.txt| in /home/hgueziri/Plus4Ibis/PlusBuild-bin/PlusLib/src/PlusServer/Tools/PlusServer.cxx(81)
|DEBUG|000.000957| Device set configuration is read from file: /home/hgueziri/.ibis/PlusToolkit/PlusConfigFiles/PlusDeviceSet_Server_Atracsys.xml| in /home/hgueziri/Plus4Ibis/PlusBuild-bin/PlusLib/src/PlusCommon/vtkPlusConfig.cxx(249)
|DEBUG|000.001047| Device set configuration file contents: 
 <PlusConfiguration version="2.7">
   <DataCollection StartupDelaySec="1.0">
     <DeviceSet Name="PlusServer: Atracsys" Description="Broadcasting tracking data from Atracsys spryTrack180, fusionTrack250 or fusionTrack500 through OpenIGTLink." />
     <Device
       Id="TrackerDevice"
       Type="AtracsysTracker"
       MaxMissingFiducials="1"
       MaxMeanRegistrationErrorMm="1.0"
       ActiveMarkerPairingTimeSec="15"
       ToolReferenceFrame="Tracker">
       <DataSources>
         <DataSource Type="Tool" Id="Stylus" TrackingType="PASSIVE" GeometryFile="/home/hgueziri/.ibis/PlusToolkit/PlusConfigFiles/AtracsysTools/101-aesculap-fs-613-pointer-factory.ini" />
         <DataSource Type="Tool" Id="Reference" TrackingType="PASSIVE" GeometryFile="/home/hgueziri/.ibis/PlusToolkit/PlusConfigFiles/AtracsysTools/pCs502x10.ini" />
       </DataSources>
       <OutputChannels>
         <OutputChannel Id="TrackerStream">
           <DataSource Type="Tool" Id="Stylus" />
           <DataSource Type="Tool" Id="Reference" />
         </OutputChannel>
       </OutputChannels>
     </Device>
   </DataCollection>
   <PlusOpenIGTLinkServer MaxNumberOfIgtlMessagesToSend="1" MaxTimeSpentWithProcessingMs="50" ListeningPort="18944" SendValidTransformsOnly="TRUE" OutputChannelId="TrackerStream">
     <DefaultClientInfo>
       <MessageTypes>
         <Message Type="TRANSFORM" />
       </MessageTypes>
       <TransformNames>
         <Transform Name="StylusToReference" />
       </TransformNames>
     </DefaultClientInfo>
   </PlusOpenIGTLinkServer>
 </PlusConfiguration>
| in /home/hgueziri/Plus4Ibis/PlusBuild-bin/PlusLib/src/PlusCommon/vtkPlusConfig.cxx(252)
|TRACE|000.001122| vtkPlusDataCollector::ReadConfiguration()| in /home/hgueziri/Plus4Ibis/PlusBuild-bin/PlusLib/src/PlusDataCollection/vtkPlusDataCollector.cxx(77)
|DEBUG|000.001159| StartupDelaySec: 1.000000| in /home/hgueziri/Plus4Ibis/PlusBuild-bin/PlusLib/src/PlusDataCollection/vtkPlusDataCollector.cxx(105)
|TRACE|000.001606| vtkPlusAtracsysTracker::vtkPlusAtracsysTracker()| in /home/hgueziri/Plus4Ibis/PlusBuild-bin/PlusLib/src/PlusDataCollection/Atracsys/vtkPlusAtracsysTracker.cxx(89)
|TRACE|000.001638| vtkPlusAtracsysTracker::ReadConfiguration| in /home/hgueziri/Plus4Ibis/PlusBuild-bin/PlusLib/src/PlusDataCollection/Atracsys/vtkPlusAtracsysTracker.cxx(114)
|TRACE|000.001667| TrackerDevice: vtkPlusDevice::ReadConfiguration| in /home/hgueziri/Plus4Ibis/PlusBuild-bin/PlusLib/src/PlusDataCollection/vtkPlusDevice.cxx(931)
|TRACE|000.004007| vtkPlusDataSource::ReadConfiguration| in /home/hgueziri/Plus4Ibis/PlusBuild-bin/PlusLib/src/PlusDataCollection/vtkPlusDataSource.cxx(253)
|DEBUG|000.004051| Buffer size is not defined in source element "StylusToTracker". Using default buffer size: 150| in /home/hgueziri/Plus4Ibis/PlusBuild-bin/PlusLib/src/PlusDataCollection/vtkPlusDataSource.cxx(391)
|DEBUG|000.004084| AveragedItemsForFiltering is not defined in source element "StylusToTracker". Using default value: 20| in /home/hgueziri/Plus4Ibis/PlusBuild-bin/PlusLib/src/PlusDataCollection/vtkPlusDataSource.cxx(401)
|TRACE|000.006259| vtkPlusDataSource::ReadConfiguration| in /home/hgueziri/Plus4Ibis/PlusBuild-bin/PlusLib/src/PlusDataCollection/vtkPlusDataSource.cxx(253)
|DEBUG|000.006291| Buffer size is not defined in source element "ReferenceToTracker". Using default buffer size: 150| in /home/hgueziri/Plus4Ibis/PlusBuild-bin/PlusLib/src/PlusDataCollection/vtkPlusDataSource.cxx(391)
|DEBUG|000.006316| AveragedItemsForFiltering is not defined in source element "ReferenceToTracker". Using default value: 20| in /home/hgueziri/Plus4Ibis/PlusBuild-bin/PlusLib/src/PlusDataCollection/vtkPlusDataSource.cxx(401)
|DEBUG|000.006361| TrackerDevice: Unable to find acquisition rate in device element when it is required, using default 30| in /home/hgueziri/Plus4Ibis/PlusBuild-bin/PlusLib/src/PlusDataCollection/vtkPlusDevice.cxx(1050)
|DEBUG|000.006624| TrackerDevice: Local time offset was not defined in device configuration| in /home/hgueziri/Plus4Ibis/PlusBuild-bin/PlusLib/src/PlusDataCollection/vtkPlusDevice.cxx(1103)
|TRACE|000.006710| vtkPlusDataCollector::GetDevice( aDevice, TrackerDevice)| in /home/hgueziri/Plus4Ibis/PlusBuild-bin/PlusLib/src/PlusDataCollection/vtkPlusDataCollector.cxx(431)
|INFO|000.006753| Server status: Reading configuration.| in /home/hgueziri/Plus4Ibis/PlusBuild-bin/PlusLib/src/PlusServer/Tools/PlusServer.cxx(92)
|DEBUG|000.006809| vtkIGSIOTransformRepository::ReadConfiguration: no CoordinateDefinitions element was found| in /home/hgueziri/Plus4Ibis/PlusBuild-bin/IGSIO/Source/IGSIOCommon/vtkIGSIOTransformRepository.cxx(627)
|INFO|000.006835| Server status: Connecting to devices.| in /home/hgueziri/Plus4Ibis/PlusBuild-bin/PlusLib/src/PlusServer/Tools/PlusServer.cxx(103)
|TRACE|000.006857| vtkPlusDataCollector::Connect()| in /home/hgueziri/Plus4Ibis/PlusBuild-bin/PlusLib/src/PlusDataCollection/vtkPlusDataCollector.cxx(359)
|DEBUG|000.006879| TrackerDevice: vtkPlusDevice::Connect: TrackerDevice| in /home/hgueziri/Plus4Ibis/PlusBuild-bin/PlusLib/src/PlusDataCollection/vtkPlusDevice.cxx(1226)
|TRACE|000.006902| vtkPlusAtracsysTracker::InternalConnect| in /home/hgueziri/Plus4Ibis/PlusBuild-bin/PlusLib/src/PlusDataCollection/Atracsys/vtkPlusAtracsysTracker.cxx(222)
|INFO|013.161600| Active marker pairing period started.| in /home/hgueziri/Plus4Ibis/PlusBuild-bin/PlusLib/src/PlusDataCollection/Atracsys/vtkPlusAtracsysTracker.cxx(289)
|INFO|028.161907| Active marker pairing period ended.| in /home/hgueziri/Plus4Ibis/PlusBuild-bin/PlusLib/src/PlusDataCollection/Atracsys/vtkPlusAtracsysTracker.cxx(294)
|TRACE|030.663636| vtkPlusDataCollector::Disconnect()| in /home/hgueziri/Plus4Ibis/PlusBuild-bin/PlusLib/src/PlusDataCollection/vtkPlusDataCollector.cxx(393)
|DEBUG|030.663740| TrackerDevice: vtkPlusDevice::Disconnect| in /home/hgueziri/Plus4Ibis/PlusBuild-bin/PlusLib/src/PlusDataCollection/vtkPlusDevice.cxx(1251)
|DEBUG|030.663828| TrackerDevice: Data source is already disconnected| in /home/hgueziri/Plus4Ibis/PlusBuild-bin/PlusLib/src/PlusDataCollection/vtkPlusDevice.cxx(1255)
|DEBUG|030.663969| vtkPlusDataCollector::Disconnect: All devices have been disconnected| in /home/hgueziri/Plus4Ibis/PlusBuild-bin/PlusLib/src/PlusDataCollection/vtkPlusDataCollector.cxx(409)
|TRACE|030.664083| vtkPlusDataCollector::SetLoopTimes| in /home/hgueziri/Plus4Ibis/PlusBuild-bin/PlusLib/src/PlusDataCollection/vtkPlusDataCollector.cxx(654)
|DEBUG|030.664161| No saved data source devices were found that use original timestamps, so synchronization of loop times is not performed| in /home/hgueziri/Plus4Ibis/PlusBuild-bin/PlusLib/src/PlusDataCollection/vtkPlusDataCollector.cxx(689)
|ERROR|030.663085| Failed to enable / disable wireless marker pairing.| in /home/hgueziri/Plus4Ibis/PlusBuild-bin/PlusLib/src/PlusDataCollection/Atracsys/vtkPlusAtracsysTracker.cxx(298)
|ERROR|030.663365| TrackerDevice: Cannot connect to device, ConnectInternal failed| in /home/hgueziri/Plus4Ibis/PlusBuild-bin/PlusLib/src/PlusDataCollection/vtkPlusDevice.cxx(1239)
|ERROR|030.663500| Unable to connect device: TrackerDevice.| in /home/hgueziri/Plus4Ibis/PlusBuild-bin/PlusLib/src/PlusDataCollection/vtkPlusDataCollector.cxx(369)
|ERROR|030.664240| Datacollector failed to connect to devices| in /home/hgueziri/Plus4Ibis/PlusBuild-bin/PlusLib/src/PlusServer/Tools/PlusServer.cxx(106)
```

**Comment by hgueziri** [comment]
also, sometimes I get a segmentation fault. Can the problem be a variable somewhere not initialized properly?

**Comment by markasselin** [comment]
@hgueziri I'm looking into this, it's a bit tricky without having access to the actual hardware to connect to. Will update when I have some ideas to test

**Pull request merged**